mmv1/third_party/terraform/services/storage/resource_storage_bucket_test.go.tmpl

```diff
@@ -1545,6 +1545,134 @@ func TestAccStorageBucket_SoftDeletePolicy(t *testing.T) {
+		CheckDestroy:             testAccStorageBucketDestroyProducer(t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccStorageBucket_basic_hns(bucketName, true),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"google_storage_bucket.bucket", "hierarchical_namespace.0.enabled", "true"),
+				),
+			},
+			{
+				ResourceName:            "google_storage_bucket.bucket",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"force_destroy"},
+			},
+			{
+				Config: testAccStorageBucket_basic_hns_with_data(bucketName, false),
+			},
+			{
+				ResourceName:            "google_storage_bucket.bucket",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"force_destroy"},
+			},
+			{
+				Config: testAccStorageBucket_uniformBucketAccessOnly(bucketName, true),
+			},
+		},
+	})
+}
+
+// testcase to create bucket with HNS disabled and
+// forcenew to recreate the bucket if HNS set to true
+func TestAccStorageBucket_basic_forcenew(t *testing.T) {
+	t.Parallel()
+
+	bucketName := acctest.TestBucketName(t)
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		CheckDestroy:             testAccStorageBucketDestroyProducer(t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccStorageBucket_basic_hns(bucketName, false),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"google_storage_bucket.bucket", "hierarchical_namespace.0.enabled", "false"),
+				),
+			},
+			{
+				ResourceName:            "google_storage_bucket.bucket",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"force_destroy"},
+			},
+			{
+				Config: testAccStorageBucket_basic_hns(bucketName, true),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"google_storage_bucket.bucket", "hierarchical_namespace.0.enabled", "true"),
+				),
+			},
+			{
+				ResourceName:            "google_storage_bucket.bucket",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"force_destroy"},
+			},
+		},
+	})
+}
+
+func TestAccStorageBucket_hns_force_destroy(t *testing.T) {
+	t.Parallel()
+
+	bucketName := acctest.TestBucketName(t)
+
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		CheckDestroy:             testAccStorageBucketDestroyProducer(t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccStorageBucket_basic_hns_with_data(bucketName, true),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckStorageBucketPutItem(t, bucketName),
```

Comment:
Thinking about this some more, shouldn't we actually add hierarchical data here? That way we'd also be testing if force_destroy can properly clear hierarchical data from the bucket.

Reply:
we don't have any terraform support yet to create folders in HNS enabled buckets

Reply:
Agree!
Even though we don't have support for HNS folders in terraform yet, We can add some hierarchical data through storage client and test force_destroy accordingly.

Reply:
I verified a manual testcase for  this,
1) create a bucket with HNS enabled and force_destroy=true with terraform
2) create folders and sub-folders which are HNS type content and upload objects in those folders manually on GCP console
3) now remove the resource in terraform and run terraform apply
4) all the files and objects got deleted and HNS folders left empty
5) The bucket failed to destroy as the bucket is not empty (empty folders are still there)

I verified there is no direct way to delete the empty folders recursively and list folders API also doesn't support order 
how to handle this case?

Reply:
added testcase for folders